Index: frontend/src/types/api.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/frontend/src/types/api.ts b/frontend/src/types/api.ts
--- a/frontend/src/types/api.ts	(revision 68addaf73c97de0c611c319d6668742d6ee1af2d)
+++ b/frontend/src/types/api.ts	(date 1763733970126)
@@ -25,6 +25,6 @@
     }
   }

-  console.log("--> Obtaining API_BASE as default /api")
-  return '/api';
+  console.log("--> Obtaining API_BASE as default /j-ast-app/api")
+  return '/j-ast-app/api';
 })();
\ No newline at end of file
